File: mzcd3-code/code/RecipesV1/PPRecipes/PPRAddRecipeIngredientViewController.m
/***
 * Excerpted from "Core Data in Objective-C, Third Edition",
 * published by The Pragmatic Bookshelf.
 * Copyrights apply to this code. It may not be used to create training material,
 * courses, books, articles, and the like. Contact us if you are in doubt.
 * We make no guarantees that this code is fit for any purpose.
 * Visit http://www.pragmaticprogrammer.com/titles/mzcd3 for more book information.
***/
#import "PPRAddRecipeIngredientViewController.h"

#import "PPRTextEditViewController.h"
#import "PPRSelectIngredientTypeViewController.h"

@interface PPRAddRecipeIngredientViewController() <NSFetchedResultsControllerDelegate>

@end

@implementation PPRAddRecipeIngredientViewController

- (void)viewWillAppear:(BOOL)animated
{
  [super viewWillAppear:animated];
  
  NSIndexPath *path = [NSIndexPath indexPathForRow:0 inSection:0];
  UITableViewCell *cell = [[self tableView] cellForRowAtIndexPath:path];
  [[cell detailTextLabel] setText:[[[self recipeIngredientMO] valueForKey:@"quantity"] stringValue]];
  
  path = [NSIndexPath indexPathForRow:1 inSection:0];
  cell = [[self tableView] cellForRowAtIndexPath:path];
  [[cell textLabel] setText:[[self recipeIngredientMO] valueForKeyPath:@"name"]];
  
  path = [NSIndexPath indexPathForRow:2 inSection:0];
  cell = [[self tableView] cellForRowAtIndexPath:path];
  [[cell textLabel] setText:[[self recipeIngredientMO] valueForKeyPath:@"unitOfMeasure"]];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
  NSString *identifier = [segue identifier];
  id viewController = [segue destinationViewController];
  
  if ([identifier isEqualToString:@"setQuantity"]) {
    [[viewController textField] setText:[[[self recipeIngredientMO] valueForKey:@"quantity"] stringValue]];
    
    [viewController setTextChangedBlock:^ BOOL (NSString *text, NSError **error) {
      NSNumberFormatter *numberFormatter = [[NSNumberFormatter alloc] init];
      
      NSNumber *quantity = [numberFormatter numberFromString:text];
      if (!quantity) {
        NSMutableDictionary *dict = [NSMutableDictionary dictionary];
        [dict setValue:@"Invalid quantity" forKey:NSLocalizedDescriptionKey];
        *error = [NSError errorWithDomain:@"PragProg" code:1123 userInfo:dict];
        return NO;
      }
      
      NSIndexPath *path = [NSIndexPath indexPathForRow:0 inSection:0];
      UITableViewCell *cell = [[self tableView] cellForRowAtIndexPath:path];
      [[cell detailTextLabel] setText:text];
      [[self recipeIngredientMO] setValue:quantity forKey:@"quantity"];
      
      return YES;
    }];
  } else if ([identifier isEqualToString:@"setType"]) {
    [[viewController textField] setText:[[self recipeIngredientMO] valueForKey:@"name"]];
    
    [viewController setTextChangedBlock:^ BOOL (NSString *text, NSError **error) {
      NSIndexPath *path = [NSIndexPath indexPathForRow:1 inSection:0];
      UITableViewCell *cell = [[self tableView] cellForRowAtIndexPath:path];
      [[cell detailTextLabel] setText:text];
      [[self recipeIngredientMO] setValue:text forKey:@"name"];
      
      return YES;
    }];
  } else if ([identifier isEqualToString:@"setUnits"]) {
    [[viewController textField] setText:[[self recipeIngredientMO] valueForKey:@"unitOfMeasure"]];
    
    [viewController setTextChangedBlock:^ BOOL (NSString *text, NSError **error) {
      NSIndexPath *path = [NSIndexPath indexPathForRow:2 inSection:0];
      UITableViewCell *cell = [[self tableView] cellForRowAtIndexPath:path];
      [[cell detailTextLabel] setText:text];
      [[self recipeIngredientMO] setValue:text forKey:@"unitOfMeasure"];
      
      return YES;
    }];
  }
}

- (IBAction)save:(id)sender;
{
  [[self navigationController] popViewControllerAnimated:YES];
}

- (IBAction)cancel:(id)sender;
{
  NSManagedObjectContext *moc = [[self recipeIngredientMO] managedObjectContext];
  [moc deleteObject:[self recipeIngredientMO]];
  [[self navigationController] popViewControllerAnimated:YES];
}

@end

File: mzswift-code/code/PPRecipes/PPRecipes/PPRAddRecipeIngredientViewController.swift
//
//  PPRAddRecipeIngredientViewController.swift
//  PPRecipes
//
//  Created by Marcus S. Zarra on 12/6/15.
//  Copyright © 2015 The Pragmatic Programmer. All rights reserved.
//

import Foundation
import UIKit
import CoreData

class PPRAddRecipeIngredientViewController: UITableViewController, NSFetchedResultsControllerDelegate {
  
  var recipeIngredientMO: NSManagedObject!
  
  func populateViewFromObject() {
    var cell = tableView.cellForRow(at: IndexPath(row: 0, section: 0))
    if let value = recipeIngredientMO.value(forKey: "quantity") {
      cell?.detailTextLabel?.text = value as? String
    } else {
      cell?.detailTextLabel?.text = "0"
    }
    cell = tableView.cellForRow(at: IndexPath(row: 1, section: 0))
    if let value = recipeIngredientMO.value(forKey: "name") {
      cell?.detailTextLabel?.text = value as? String
    } else {
      cell?.detailTextLabel?.text = ""
    }
    cell = tableView.cellForRow(at: IndexPath(row: 2, section: 0))
    if let value = recipeIngredientMO.value(forKey: "unitOfMeasure") {
      cell?.detailTextLabel?.text = value as? String
    } else {
      cell?.detailTextLabel?.text = ""
    }
  }
  
  override func viewWillAppear(_ animated: Bool) {
    super.viewWillAppear(animated)
    populateViewFromObject()
  }
  
  override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
    guard let identifier = segue.identifier else { fatalError("Segue without an identifier") }
    guard let controller = segue.destination as? PPRTextEditViewController else { fatalError("Wrong destination view controller") }
    
    switch identifier {
    case "setQuantity":
      let quantity = recipeIngredientMO.value(forKey: "quantity") as! NSNumber
      controller.text = quantity.stringValue
      controller.textChangedClosure = {
        (newText: String) in
        let numberFormatter = NumberFormatter()
        guard let quantity = numberFormatter.number(from: newText) else {
          let userInfo = [NSLocalizedDescriptionKey: "Invalid quantity"]
          throw NSError(domain: "PragProg", code: 1123, userInfo: userInfo)
        }
        self.recipeIngredientMO.setValue(quantity, forKey: "quantity")
      }
    case "setType":
      let type = recipeIngredientMO.value(forKey: "name") as! String
      controller.text = type
      controller.textChangedClosure = {
        (newText: String) in
        self.recipeIngredientMO.setValue(type, forKey: "name")
      }
    case "setUnits":
      let type = recipeIngredientMO.value(forKey: "unitOfMeasure") as! String
      controller.text = type
      controller.textChangedClosure = {
        (newText: String) in
        self.recipeIngredientMO.setValue(type, forKey: "unitOfMeasure")
      }
    default:
      print("Unknown identifier: \(identifier)")
    }
    populateViewFromObject()
  }
  
  func save(_ sender: AnyObject) {
    navigationController!.popViewController(animated: true)
  }
  
  func cancel(_ sender: AnyObject) {
    let moc = recipeIngredientMO.managedObjectContext
    moc?.delete(recipeIngredientMO)
    navigationController!.popViewController(animated: true)
  }
}
